protocol/vmutil/builder.go

```diff
@@ -1,31 +1,111 @@
 package vmutil
 
-import "chain/protocol/vm"
+import (
+	"encoding/binary"
+	"fmt"
+
+	"chain/protocol/vm"
+)
 
 type Builder struct {
-	Program []byte
+	program     []byte
+	jumpCounter int
+
+	// Maps a jump target number to its absolute address.
+	jumpAddr map[int]uint32
+
+	// Maps a jump target number to the list of places where its
+	// absolute address must be filled in once known.
+	jumpPlaceholders map[int][]int
 }
 
 func NewBuilder() *Builder {
-	return &Builder{}
+	return &Builder{
+		jumpAddr:         make(map[int]uint32),
+		jumpPlaceholders: make(map[int][]int),
+	}
 }
 
+// AddInt64 adds a pushdata instruction for an integer value.
 func (b *Builder) AddInt64(n int64) *Builder {
-	b.Program = append(b.Program, vm.PushdataInt64(n)...)
+	b.program = append(b.program, vm.PushdataInt64(n)...)
 	return b
 }
 
+// AddData adds a pushdata instruction for a given byte string.
 func (b *Builder) AddData(data []byte) *Builder {
-	b.Program = append(b.Program, vm.PushdataBytes(data)...)
+	b.program = append(b.program, vm.PushdataBytes(data)...)
 	return b
 }
 
+// AddRawBytes simply appends the given bytes to the program. (It does
+// not introduce a pushdata opcode.)
 func (b *Builder) AddRawBytes(data []byte) *Builder {
-	b.Program = append(b.Program, data...)
+	b.program = append(b.program, data...)
 	return b
 }
 
+// AddOp adds the given opcode to the program.
 func (b *Builder) AddOp(op vm.Op) *Builder {
-	b.Program = append(b.Program, byte(op))
+	b.program = append(b.program, byte(op))
 	return b
 }
+
+// NewJumpTarget allocates a number that can be used as a jump target
+// in AddJump and AddJumpIf. Call SetJumpTarget to associate the
+// number with a program location.
+func (b *Builder) NewJumpTarget() int {
+	b.jumpCounter++
+	return b.jumpCounter
+}
+
+// AddJump adds a JUMP opcode whose target is the given target
+// number. The actual program location of the target does not need to
+// be known yet, as long as SetJumpTarget is called before Build.
+func (b *Builder) AddJump(target int) *Builder {
+	return b.addJump(vm.OP_JUMP, target)
+}
+
+// AddJump adds a JUMPIF opcode whose target is the given target
+// number. The actual program location of the target does not need to
+// be known yet, as long as SetJumpTarget is called before Build.
+func (b *Builder) AddJumpIf(target int) *Builder {
+	return b.addJump(vm.OP_JUMPIF, target)
+}
+
+func (b *Builder) addJump(op vm.Op, target int) *Builder {
+	b.AddOp(op)
+	var addrBytes [4]byte
+	if addr, ok := b.jumpAddr[target]; ok {
```

Comment:
We can reduce code paths by removing the conditional here and always doing the else block.